src/mqttgw_dyn_srv.erl

```diff
@@ -0,0 +1,251 @@
+            delete_dynsub(Subject, Data),
+            {CorrData, RespTopic, SenderConn, SenderId,
+                UniqueId, SessionPairId, Time} = DynsubRespData,
+
+            %% Send a response to the application.
+            mqttgw:send_dynsub_response(
+                CorrData, RespTopic, SenderConn, SenderId,
+                UniqueId, SessionPairId, Time),
+
+            %% Send a multicast event to the application
+            mqttgw:send_dynsub_multicast_event(
+                <<"subscription.delete">>, Subject, Data, SenderConn, SenderId,
+                UniqueId, SessionPairId, Time),
+            ok;
+        %% Subject is on a different node, we need to delete the dynsub on that node.
+        [{Node, _, _Subs}] ->
+            Topic = authz_subscription_topic(Data),
+            remove_dynsub(Node, Subject, Topic, Data, DynsubRespData),
+            ok
+    end.
+
+start_dynsub(Node, Subject, Topic, Data, DynsubRespData) ->
+    gen_server:cast(
+        {global, {?MODULE, Node}},
+        {start_dynsub, Subject, Topic, Data, DynsubRespData, node()}
+    ).
+
+register_dynsub_in_progress(Node, Subject, Topic, DynsubRespData) ->
+    gen_server:cast(
+        {global, {?MODULE, Node}},
+        {register_dynsub_in_progress, Subject, Topic, DynsubRespData}
+    ).
+
+remove_dynsub(Node, Subject, Topic, Data, DynsubRespData) ->
+    gen_server:cast(
+        {global, {?MODULE, Node}},
+        {remove_dynsub, Subject, Topic, Data, DynsubRespData, node()}
+    ).
+
+register_dynsub_removal_in_progress(Node, Subject, Topic, Data, DynsubRespData) ->
+    gen_server:cast(
+        {global, {?MODULE, Node}},
+        {register_dynsub_removal_in_progress, Subject, Topic, Data, DynsubRespData}
+    ).
+
+-spec authz_subscription_topic(mqttgw_dynsub:data()) -> mqttgw:topic().
+authz_subscription_topic(Data) ->
+    #{app := App,
+      object := Object,
+      version := Version} = Data,
+    [<<"apps">>, App, <<"api">>, Version | Object].
+
+-spec create_sub(mqttgw_dynsub:subject(), mqttgw_dynsub:data()) -> ok.
+create_sub(Subject, Data) ->
+    QoS = 1,
+    Topic = authz_subscription_topic(Data),
+    mqttgw_broker:subscribe(Subject, [{Topic, QoS}]),
+
+    error_logger:info_msg(
+        "Dynamic subscription: ~p has been created "
+        "for the subject = '~s'",
+        [Data, Subject]),
+
+    ok.
+
+-spec delete_dynsub(mqttgw_dynsub:subject(), mqttgw_dynsub:data()) -> ok.
+delete_dynsub(Subject, Data) ->
+    Topic = authz_subscription_topic(Data),
+    mqttgw_broker:unsubscribe(Subject, [Topic]),
+
+    error_logger:info_msg(
+        "Dynamic subscription: ~p has been deleted "
+        "for the subject = '~s'",
+        [Data, Subject]),
+
+    ok.
+
+init([]) ->
+    EHandler = vmq_subscriber_db:subscribe_db_events(),
+    State = #{event_handler => EHandler, subs_queue => #{}, removals_queue => #{}},
+    {ok, State}.
+
+handle_cast({start_dynsub, Subject, Topic, Data, DynsubRespData, From}, State) ->
```

Comment:
`register_dynsub` may be a better name for `start_dynsub`, since it's an initial step before `register_dynsub_in_progress`. I forgot to mention this last time )